cpp/terrain/terrain_search.cpp

```diff
@@ -0,0 +1,80 @@
+// Copyright 2015-2015 the openage authors. See copying.md for legal info.
+
+#include <cmath>
+
+#include "terrain.h"
+#include "terrain_object.h"
+#include "terrain_search.h"
+
+namespace openage {
+
+std::shared_ptr<TerrainObject> find_near(const TerrainObject &start, 
+                                         std::function<bool(const TerrainObject &)> found) {
+	
+	auto terrain = start.get_terrain();
+	auto tile = start.pos.draw.to_tile3().to_tile();
+	TerrainSearch search(terrain, tile);
+
+	for (int i = 0; i < 500; ++i) {
+		for (auto o : terrain->get_data(tile)->obj) {
+			auto shared_o = o.lock();
+			if (found(*shared_o)) {
+				return shared_o;
+			}
+		}
+		tile = search.next_tile();
+	}
+
+	return nullptr;
+}	
+
+TerrainSearch::TerrainSearch(std::shared_ptr<Terrain> t, coord::tile s)
+	:
+	TerrainSearch(t, s, .0f) {
```

Comment:
not an important issue, but I think the closing brace should be in the same line.